mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxACL.java

```diff
@@ -256,96 +731,142 @@
+            .stream()
+            .map(entry -> Pair.of(entry.getKey(),
+                Optional.ofNullable(other.getEntries().get(entry.getKey()))
+                    .map(Throwing.function(exceptValue -> entry.getValue().except(exceptValue)))
+                    .orElse(entry.getValue())))
+            .filter(pair -> !pair.getValue().isEmpty())
+            .collect(Guavate.toImmutableMap(Map.Entry::getKey, Map.Entry::getValue)));
+    }
 
-    /**
-     * TODO except.
-     * 
-     * @param key
-     * @param toRemove
-     * @return
-     * @throws UnsupportedRightException
-     */
-    MailboxACL except(MailboxACLEntryKey key, MailboxACLRights toRemove) throws UnsupportedRightException;
+    public MailboxACL except(EntryKey key, Rfc4314Rights mailboxACLRights) throws UnsupportedRightException {
+        return except(new MailboxACL(new Entry(key, mailboxACLRights)));
+    }
 
     /**
      * {@link Map} of entries.
-     * 
+     *
      * @return the entries.
      */
-    Map<MailboxACLEntryKey, MailboxACLRights> getEntries();
+    public Map<EntryKey, Rfc4314Rights> getEntries() {
+        return entries;
+    }
 
     /**
      * Replaces the entry corresponding to the given {@code key} with
      * {@code toAdd}link MailboxACLRights}.
-     * 
+     *
      * Implementations must return a new unmodifiable instance of
      * {@link MailboxACL}. However, implementations may decide to return this in
      * case the result would be equal to it.
-     * 
+     *
      * Implementations must ensure that the result does not contain entries with
      * empty rigths. E.g. "user1:lr;user2:lrwt".replace("user1",
      * MailboxACLRights.EMPTY) should return "user2:lrwt" rather than
      * "user1:;user2:lrwt". The same result should be returned by
      * "user1:lr;user2:lrwt".replace("user1", null).
-     * 
+     *
      * @param key
      * @param toAdd
      * @return
      * @throws UnsupportedRightException
      */
-    MailboxACL replace(MailboxACLEntryKey key, MailboxACLRights toAdd) throws UnsupportedRightException;
+    public MailboxACL replace(EntryKey key, Rfc4314Rights replacement) throws UnsupportedRightException {
+        if (entries.containsKey(key)) {
+            return new MailboxACL(
+                entries.entrySet()
+                    .stream()
+                    .map(entry -> Pair.of(entry.getKey(),
+                        entry.getKey().equals(key) ? replacement : entry.getValue()))
+                    .filter(pair -> pair.getValue() != null && !pair.getValue().isEmpty())
+                    .collect(Guavate.toImmutableMap(Pair::getKey, Pair::getValue)));
+        } else {
+            return new MailboxACL(
+                ImmutableMap.<EntryKey, Rfc4314Rights>builder()
+                    .putAll(entries)
+                    .put(key, replacement)
+                    .build());
+        }
+    }
+
+    public String toString() {
+        return entries == null ? "" : entries.toString();
```

Comment:
toString returning en empty string is strange

Reply:
Let's use MoreObjects ;-)